From 00d4d7623f48e81dc3005c1faf96051cdac677e2 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:06:05 -0400
Subject: [ethernet] qlcnic: Correct off-by-one errors in bounds checks

Message-id: <1408626422-3941-82-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88624
O-Subject: [PATCH RHEL 7.1 081/138] qlcnic: Correct off-by-one errors in bounds checks
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Manish Chopra <manish.chopra@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id 462bed4870a813bb3ab33254de70757a92d0dc69

o Bound checks should be >= instead of > for number of receive descriptors
  and number of receive rings.

Signed-off-by: Manish Chopra <manish.chopra@qlogic.com>
Signed-off-by: Jitendra Kalsaria <jitendra.kalsaria@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_io.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_io.c
index 9469cf6..0ae31f4 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_io.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_io.c
@@ -1149,13 +1149,13 @@ qlcnic_process_lro(struct qlcnic_adapter *adapter,
  u16 lro_length, length, data_offset, t_vid, vid = 0xffff;
  u32 seq_number;
 
- if (unlikely(ring > adapter->max_rds_rings))
+ if (unlikely(ring >= adapter->max_rds_rings))
   return NULL;
 
  rds_ring = &recv_ctx->rds_rings[ring];
 
  index = qlcnic_get_lro_sts_refhandle(sts_data0);
- if (unlikely(index > rds_ring->num_desc))
+ if (unlikely(index >= rds_ring->num_desc))
   return NULL;
 
  buffer = &rds_ring->rx_buf_arr[index];
@@ -1661,13 +1661,13 @@ qlcnic_83xx_process_lro(struct qlcnic_adapter *adapter,
  u16 vid = 0xffff;
  int err;
 
- if (unlikely(ring > adapter->max_rds_rings))
+ if (unlikely(ring >= adapter->max_rds_rings))
   return NULL;
 
  rds_ring = &recv_ctx->rds_rings[ring];
 
  index = qlcnic_83xx_hndl(sts_data[0]);
- if (unlikely(index > rds_ring->num_desc))
+ if (unlikely(index >= rds_ring->num_desc))
   return NULL;
 
  buffer = &rds_ring->rx_buf_arr[index];
-- 
1.7.1